### ISSUE TITLE
Component to create: datepicker

### ISSUE THREAD

**Issue by andrewconnell** [ngOfficeUIFabric/ng-officeuifabric#19]
Official link to component description & demo: http://dev.office.com/fabric/components/datepicker

_Screenshot of web analytics for the original http://officeuifabric.com site by @andrewconnell. Use this to see what are the most popular components: http://imgur.com/gallery/buUcyQv_

**Comment by andrewconnell** [comment]
Started by @rolandoldengarm...

**Comment by ghost** [comment]
The Office UI Fabric component uses pickadate.js which requires jQuery. Our rule is that we don't have jQuery dependencies.
If I would write this from scratch, this would be quite time consuming. Does anyone have tips, e.g. another datepicker I can use as a reference? All tips welcome.

**Comment by andrewconnell** [comment]
This one bums me out... to the best of my understanding, this is the only one that has a tight coupling on jQuery. I agree... it's too much at this stage to rewrite. I'm personally OK with making this exception saying we can tell the user to have a dependency on jQuery... maybe one day someone can rewrite the whole thing to strip jQuery... but not today. Maybe with inspiration from https://material.angularjs.org/latest/demo/datepicker.

In reality we have two dependencies: `pickadate.js` & `jquery.js`. So how to handle this. A few options:
- [ ] we include both in our webpack bundle so the user doesn't need to do anything other than include our library... but if they already have jQuery on the page, we now have two copies (not ideal)
- [ ] we include `pickadate.js` and doc that they need a jQuery reference on the page
- [ ] we include `pickadate.js` & directive checks if jQuery on the page... if not, CDN reference is added dynamically
- [ ] we include neither, directive checks if both are on the page and if not, CDN reference is added dynamically

I personally prefer the last option. This way a developer doesn't have to do any extra work but if they are in an environment sans internet access, they can manually add their own references and it will work.

Either way, I think we doc this clearly in the directive docs and in the demo page.

Thoughts?

**Comment by jjczopek** [comment]
**TL;DR;**
Option 4 from @andrewconnell sounds more reasonable;

**Long story**
Regarding first option, we definitely don't want just to include jQuery in our library, because we will likely create conflicts with other jQuery versions that are on the page (which can happen). On the other hand we could ship it with jQuery, but hide _our_ jQuery in other namespace, e.g. uifjq.

As for other options, I think the last one is quite nice. By default developer doesn't have to take care about dependecies and we can just reference from CDN when needed. In case the app is supposed to be run on environment without access to Internet, they are supposed to provide those libraries according to their preference.

**Comment by ghost** [comment]
Yes, I agree with option 4 too!

**Comment by s-KaiNet** [comment]
Option 4 may be a tricky under some circumstances. For example what if jquery is loaded dynamically, or by `require`-ing, in that case on a moment of checking, window.jQuery may be null....we will include our version and after that page's version will be loaded.  
I'm not sure how to deal with that. May be take a look at option 5 - add dependency on angular's pickadate implementation - [angular-pickadate](https://github.com/restorando/angular-pickadate).  
As fabric depends on jquery component, why can't we depend on angular implementation?

**Comment by andrewconnell** [comment]
oooo... option 5 warrants more investigation...

**Comment by ghost** [comment]
I think it is cleaner to depend on another Angular module instead of jQuery entirely. However, it will obviously take (much) more time as I cannot reuse any of the scripts.
But I'd like to go down the path of using the angular pickadate implementation as that is much cleaner than depending on jQuery entirely plus an additional library pickadate.js
How do we implement dependencies on other Angular libraries? Use a similar approach as outlined above re jQuery? Or just include the JS in the pickadate directive I'm going to build? I think the latter makes the most sense. The issue with depending on the dev to include it from angular-pickadate cdn is that pickadate may release a new version which is incompatible with our directive.
By including the JS we're sure that it works.

**Comment by s-KaiNet** [comment]
I'm agreed with you on point to depends on ng-pickadate. How I see all this - you need to add dependency in npm on ng-pickadate, then in your datepicker.ts in the top you need to explicitly require .js and .css from `node_modules/ngpickadate/dist/...`. Something like this (didn't test): 

``` javascript
require('./../../node_modules/ngpickadate/dist/pickadate.js')
require('./../../node_modules/ngpickadate/dist/pickadate.css')
```

To load css you need to add css loader for webpack. In that way webpack will include pickadate only if your module is using, also there is no problem with cdn, because our package will contain pickadate inside.

**Comment by andrewconnell** [comment]
Agreed on depending on another Angular implementation. The fun part will getting the styles just right.

Now for shipping the resources... let's think about our options here. Today we aren't including any external dependencies (Angular JS, CSS for Fabric) because we aren't tightly coupled. But for something like this... does it make sense to do that? For instance, we look into configuring our webpack build to injest the JS & CSS into the bundle & include something in the comment banner that gives credit to the source... that would make things easier for consumers... but it's a slippery slope.

I think for now, we simply _reference_ external JS & CSS and provide guidance. Today, demos that use the `uif-progress-indicator` have to [include a separate CSS file](https://github.com/ngOfficeUIFabric/ng-officeuifabric/blob/master/src/components/progressindicator/demo/index.html#L11) because of a bug in fabric 1.2.1 [#282](https://github.com/OfficeDev/Office-UI-Fabric/pull/282).

We just have to provide the proper docs & guidance.

**Comment by s-KaiNet** [comment]
Good point. I'm thinking primary from the consumer side - include script in the page and forget :) Also UI Fabric includes pickadate in /dist file.  
But in terms of loosely coupling that's not good. As we don't want tightly coupled components, then we definitely don't need to include this.

**Comment by ghost** [comment]
I've had a look at the angular-pickadate. While I like it is a non-jQuery solution, I see some issues:
- There is no NPM package (or any other package as far as I could see)
- Not much activity, e.g. demo does not work
- The resulting HTML will be different, so I'll have to do a lot of work to make it work with Fabric CSS.

I think it is possible, but I think we should postpone a non-jQuery solution until after release 1.0.0. Right now it's just out of reach as it will take a lot of time.

So my plan of attack is to include the pickadate npm package. I can make this a dev dependency, so everyone using our NPM package doesn't have to do anything. Anyone else has to install the pickadate npm package. Nothing more. What about that?

**Comment by andrewconnell** [comment]
I think that’s fair… that will be the exception to everything we’re doing. It sounds like the Office UI Fabric team is taking the same approach for now…

https://github.com/OfficeDev/Office-UI-Fabric/issues/222#issuecomment-175843914

**Comment by ghost** [comment]
Ok cool. Next discussion point. The test scripts will require datepicker library to be included. 
I tried to require it in the specs file, but that fails (which seems an issue with pickadate). It was fixed when I applied this change: https://github.com/amsul/pickadate.js/pull/779

Easiest would be to include it in the tests file, but then it's always loaded. Right now that's not an issue, as we always run all tests, but in the future that may change.

**Comment by andrewconnell** [comment]
I think it's fine to load it the same way we are loading jQuery for tests. Seems like it would go right after this line: https://github.com/ngOfficeUIFabric/ng-officeuifabric/blob/dev/config/karma.ts#L48
